scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetListFormatRule.scala

```diff
@@ -0,0 +1,262 @@
+package com.twitter.scalding.parquet.scrooge
+
+import java.util
+
+import org.apache.parquet.schema.{GroupType, OriginalType, PrimitiveType, Type}
+import org.slf4j.LoggerFactory
+
+/**
+ * Rule to convert parquet schema of legacy list type to standard one
+ * namely 3-level list structure as recommended in
+ * https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#lists
+ *
+ * More specifically this handles converting from parquet file created by
+ * {{@code org.apache.parquet.thrift.ThriftSchemaConvertVisitor}} which always suffix
+ * list element with "_tuple".
+ */
+private[scrooge] object ParquetListFormatRule extends ParquetCollectionFormatRule {
+
+  private val LOGGER = LoggerFactory.getLogger(getClass)
+
+  def formatForwardCompatibleRepeatedType(repeatedSourceType: Type, repeatedTargetType: Type) = {
+    (
+      findFirstListRule(repeatedSourceType, Source),
+      findFirstListRule(repeatedTargetType, Target)
+    ) match {
+      case (Some(sourceRule), Some(targetRule)) => {
+        if (sourceRule == targetRule) {
+          repeatedSourceType
+        } else {
+          val elementType = sourceRule.elementType(repeatedSourceType)
+          targetRule.createCompliantRepeatedType(
+            typ = elementType,
+            name = elementType.getName,
+            isElementRequired = sourceRule.isElementRequired(repeatedSourceType),
+            originalType = sourceRule.elementOriginalType(repeatedSourceType)
+          )
+        }
+      }
+      case _ => repeatedSourceType
+    }
+  }
+
+  def isGroupList(projection: Type): Boolean = {
+    if (projection.isPrimitive) {
+      false
+    } else {
+      val groupProjection = projection.asGroupType
+      groupProjection.getOriginalType == OriginalType.LIST &&
+        groupProjection.getFieldCount == 1 &&
+        groupProjection.getFields.get(0).isRepetition(Type.Repetition.REPEATED)
+    }
+
+  }
+
+  def findFirstListRule(repeatedType: Type,
+                        sourceOrTarget: SourceOrTarget): Option[ParquetListFormatRule] = {
+    val ruleFound = sourceOrTarget.rules.find(rule => rule.check(repeatedType))
+    if (ruleFound.isEmpty) LOGGER.warn(s"Unable to find matching rule for ${sourceOrTarget.name} schema:\n$repeatedType")
+    ruleFound
+  }
+
+  def wrapElementAsRepeatedType(rule: ParquetListFormatRule, repeatedType: Type, elementType: Type): Type = rule.createCompliantRepeatedType(
+    elementType,
+    rule.elementName(repeatedType),
+    // if repeated or required, it is required
+    !elementType.isRepetition(Type.Repetition.OPTIONAL),
+    elementType.getOriginalType)
+}
+
+/**
+ * Helper to specify supported source/target conversion.
+ */
+private[scrooge] sealed trait SourceOrTarget {
+  def rules: Seq[ParquetListFormatRule]
+  def name: String
+}
+
+private[scrooge] object Source extends SourceOrTarget {
```

Comment:
what's the advantage of distinguishing between Source and Target? is it so we can enforce a directionality in terms of parquet spec conversion? I'm wondering if we're being too specific here and wind up with a direction that we didn't engineer for.

Reply:
I guess since scalding is going to be stuck on the older parquet spec, this would not happen.

Reply:
ack. Will see how I can do end-to-end testing in this repo.